File: 15-3sum/15-3sum.js
/**
 * @param {number[]} nums
 * @return {number[][]}
 */
var threeSum = function(nums) {
    let res = []
    nums = nums.sort((a, b) => a - b)
    for (let i = 0; i < nums.length - 2; i++) {
        let j = i + 1
        let k = nums.length - 1
        if (i > 0 && nums[i] === nums[i - 1]) continue
        while (j < k) {
            const sum = nums[i] + nums[j] + nums[k]
            if (sum === 0) {
                res.push([nums[i] ,nums[j] , nums[k]])
                j++
                while (nums[j] === nums[j - 1] && j < k)j++
            } else if (sum > 0) {
                k--
            } else {
                j++
            }
        }
    }
    return res
};

File: 15-3sum/15-3sum.py
class Solution(object):
    def threeSum(self, nums):
        """
        :type nums: List[int]
        :rtype: List[List[int]]
        """
        res = []
        nums.sort()
        for i in range(len(nums) - 2):
            if i > 0 and nums[i] == nums[i - 1]:
                continue
            j = i + 1
            k = len(nums) - 1
            
            while j < k:
                sum = nums[i] + nums[j] + nums[k]
                if sum == 0:
                    res.append((nums[i], nums[j], nums[k]))
                    while j < k and nums[j] == nums[j+1]:
                        j += 1
                    while j < k and nums[k] == nums[k-1]:
                        k -= 1
                    j += 1; k -= 1
                elif sum > 0:
                    k -= 1
                else: 
                    j += 1
        return res
        